persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java

```diff
@@ -94,37 +114,179 @@ public void beforeChange(
+    for (EventListener listener : listeners) {
+      listener.onLeaveCluster(nativeAddress, EventListener.NO_PORT);
+    }
+  }
 
-    for (IEndpointLifecycleSubscriber subscriber : subscribers) subscriber.onLeaveCluster(endpoint);
+  private void notifyJoinCluster(InetAddress endpoint) {
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onJoinCluster(nativeAddress, EventListener.NO_PORT);
+    }
   }
 
-  @Override
-  public void onRestart(InetAddress endpoint, EndpointState state) {}
+  private void notifyRpcChange(InetAddress endpoint, boolean isReady) {
+    if (isReady) {
+      notifyUp(endpoint);
+    } else {
+      notifyDown(endpoint);
+    }
+  }
 
-  private void addStargateNode(InetAddress endpoint, EndpointState epState) {
-    if (!isStargateNode(epState) || !liveStargateNodes.add(endpoint)) {
+  private void notifyUp(InetAddress endpoint) {
+    if (!StorageService.instance.isRpcReady(endpoint) || !Gossiper.instance.isAlive(endpoint)) {
       return;
     }
 
-    for (IEndpointLifecycleSubscriber subscriber : subscribers) subscriber.onJoinCluster(endpoint);
+    if (endpointsPendingJoinedNotification.remove(endpoint)) {
+      notifyJoinCluster(endpoint);
+    }
+
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onUp(nativeAddress, EventListener.NO_PORT);
+    }
+  }
+
+  private void applyState(
+      InetAddress endpoint, ApplicationState state, VersionedValue value, EndpointState epState) {
+    final ExecutorService executor = StageManager.getStage(Stage.MUTATION);
+    switch (state) {
+      case RELEASE_VERSION:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "release_version", value.value, executor);
+        break;
+      case DC:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "data_center", value.value, executor);
+        break;
+      case RACK:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "rack", value.value, executor);
+        break;
+      case RPC_ADDRESS:
+        try {
+          StargateSystemKeyspace.updatePeerInfo(
+              endpoint, "rpc_address", InetAddress.getByName(value.value), executor);
+        } catch (UnknownHostException e) {
+          throw new RuntimeException(e);
+        }
+        break;
+      case SCHEMA:
+        // Use a fix schema version for all peers (always in agreement) because stargate waits
+        // for DDL queries to reach agreement before returning.
+        StargateSystemKeyspace.updatePeerInfo(
+            endpoint, "schema_version", StargateSystemKeyspace.SCHEMA_VERSION, executor);
+
+        // This fix schedules a schema pull for the non-member node and is required because
+        // `StorageService.onChange()` doesn't do this for non-member nodes.
+        MigrationManager.instance.scheduleSchemaPull(endpoint, epState);
```

Comment:
I see that `updatePeerInfo` returns `Future` and we are not blocking on wait for a previous step. Isn't there a risk for race condition here? Shouldn't we chain the `MigrationManager.instance.scheduleSchemaPull(endpoint, epState);` when the Future returned by the `updatePeerInfo` completes?

Reply:
I thought about this too. I'm replicating what's done in `StorageService`, but maybe that's a long standing bug too? I'll add it, but I'm not sure of the implications of blocking in the gossiper thread...

>

Reply:
I don't mean blocking, just chain the subsequent operation.
If we would use `CompletableFuture` we could do smth like
 ```
CompletableFuture.from(() -> StargateSystemKeyspace.updatePeerInfo(
            endpoint, "schema_version", StargateSystemKeyspace.SCHEMA_VERSION, executor);
).thenApply( () -> MigrationManager.instance.scheduleSchemaPull(endpoint, epState));
```

Reply:
Understood. The schema version is constant and not related to pulling the schema from other endpoints. Why do these need to be chained?